src/goto-programs/goto_contractor.cpp

```diff
@@ -0,0 +1,359 @@
+//
+// Created by Mohannad Aldughaim on 09/01/2022.
+//
+
+#include "goto_contractor.h"
+
+using namespace std;
+
+
+void goto_contractort::get_constraints(goto_functionst goto_functions)
+{
+  auto function = goto_functions.function_map.find("c:@F@main");
+  auto it = function->second.body.instructions.begin();
+  while(it != function->second.body.instructions.end())
+  {
+    if(it->is_assert())
+      constraint = create_constraint_from_expr2t(it->guard);
+
+    it++;
+  }
+}
+
+void goto_contractort::get_intervals(goto_functionst goto_functions)
+{
+  //only main here
+  auto function = goto_functions.function_map.find("c:@F@main");
+  std::cout << "\n parsing assumes\nfound main: " << function->first << "\n";
+  auto it = function->second.body.instructions.begin();
+
+  while(it != function->second.body.instructions.end())
+  {
+    if(it->is_assume())
+      parse_intervals(it->guard);
+
+    it++;
+  }
+}
+
+void goto_contractort::parse_intervals(irep_container<expr2t> expr)
+{
+  symbol2tc symbol;
+  double value;
+
+  if(!is_comp_expr(expr))
+    return;
+
+  std::shared_ptr<relation_data> rel;
+  rel = dynamic_pointer_cast<relation_data>(expr);
+
+  //expr->dump();
+  //side1 1 is a symbol or typecast to symbol
+  auto side1 = rel->side_1;
+  //side 2 is always a number.
+  auto side2 = rel->side_2;
+
+  if(is_typecast2t(side1))
+    if(is_symbol2t(to_typecast2t(side1).from))
+      symbol = to_symbol2t(to_typecast2t(side1).from);
+    else
+      return;
+  else if(is_symbol2t(side1))
+    symbol = to_symbol2t(side1);
+  else
+    return;
+
+  if(!is_constant_int2t(side2) && !is_constant_floatbv2t(side2))
+    return;
+  if(is_constant_int2t(side2))
+    value = to_constant_int2t(side2).as_long();
+  else
+    return;
+
+  int index =
+    map->find((string)symbol->get_symbol_name().c_str());
+  if(index == -1) return;
+  if(is_greaterthan2t(expr) || is_greaterthanequal2t(expr))
```

Comment:
You can replace this with a switch over the expr type.

Reply:
done.